core/src/main/scala/org/apache/spark/util/ListenerBus.scala

```diff
@@ -17,53 +17,222 @@
+    new ConcurrentHashMap[String, ListenerEventExecutor[L]] ()
+
+  // Indicate if `start()` is called
+  private val started = new AtomicBoolean(false)
+  // Indicate if `stop()` is called
+  private val stopped = new AtomicBoolean(false)
 
   /**
    * Add a listener to listen events. This method is thread-safe and can be called in any thread.
    */
-  final def addListener(listener: L): Unit = {
-    listeners.add(listener)
+  final def addListener(
+    listener: L, eventListenerGroup: String = ListenerEventExecutor.DefaultEventListenerGroup):
+  Unit = synchronized {
+    var listenerEventExecutor = eventGroupToEventExecutors.get(eventListenerGroup)
+    if (listenerEventExecutor == null) {
+      listenerEventExecutor =
+        new ListenerEventExecutor[L](listener.getClass.getName, eventQueueSize)
+      eventGroupToEventExecutors.put(eventListenerGroup, listenerEventExecutor)
+
+    }
+    listenerEventExecutor.addListener(listener)
+    if (started.get()) {
+      listenerEventExecutor.start
+    }
   }
 
   /**
    * Remove a listener and it won't receive any events. This method is thread-safe and can be called
    * in any thread.
    */
   final def removeListener(listener: L): Unit = {
-    listeners.remove(listener)
+    for (eventExecutor <- eventGroupToEventExecutors.values().asScala) {
+      eventExecutor.removeListener(listener)
+    }
+  }
+
+   /**
+    * For testing only. Returns whether there is any event pending to be processed by
+    * any of the existing listener
+    */
+  def isListenerBusEmpty: Boolean = {
+    for (eventExecutor <- eventGroupToEventExecutors.values().asScala) {
+      if (!eventExecutor.isEmpty) {
+        return false
+      }
+    }
+    true
   }
 
   /**
-   * Post the event to all registered listeners. The `postToAll` caller should guarantee calling
-   * `postToAll` in the same thread for all events.
+   * Posts the event to all registered listeners. This is an async call and it does not
+   * processes the event itself. Processing of the event is done in a separate thread in
+   * the {@link ListenerEventExecutor}.
    */
   final def postToAll(event: E): Unit = {
-    // JavaConverters can create a JIterableWrapper if we use asScala.
-    // However, this method will be called frequently. To avoid the wrapper cost, here we use
-    // Java Iterator directly.
-    val iter = listeners.iterator
-    while (iter.hasNext) {
-      val listener = iter.next()
-      try {
-        doPostEvent(listener, event)
-      } catch {
-        case NonFatal(e) =>
-          logError(s"Listener ${Utils.getFormattedClassName(listener)} threw an exception", e)
+    for (listenerEventProcessor <- eventGroupToEventExecutors.values().asScala) {
+      // JavaConverters can create a JIterableWrapper if we use asScala.
+      // However, this method will be called frequently. To avoid the wrapper cost, here we use
+      // Java Iterator directly.
+      val iter = listenerEventProcessor.listeners.iterator()
```

Comment:
It feels like this logic should be in the `submit()` implementation. Right now you're peeking into its internals, and actually wrapping a runnable inside another runnable, which feels hacky.

Instead, if you just `submit()` the event itself, the task you submit to the executor can iterate over the listeners in the group, and you don't need to expose internal state of the dispatcher to this code.